app/js/enhancedDate.js

```diff
@@ -1,5 +1,239 @@
-/* 
- * add your module definition for enhancedDate here
- * assign the module to a global variable
- * /
+/*
+ * Homework 
+ * JSCRIPT300-Spring2015/Module_2
+ * by Diane Zevenbergen
+ *
+ * Note: I assumed that the setDate method was supposed to be flexible and store 'theDate' either
+ * as a Date object OR in epoch milliseconds, depending on the argument that was passed.
+ * So wherever I use a Date object method such as getTime(), I call a fcn getDateObj to
+ * convert 'theDate' if needed.
+ */
 
+var enhancedDate = (function iife() {
+    
+    var theDate;
+    
+    /*
+     * This function checks whether 'theDate' has been set by the setDate
+     * method yet.  If not, it calls setDate & sets date to now.
+     * Called  by all methods in publicAPI (except setDate).
+     */
+    function verifyDate(){
+        if (typeof theDate === 'undefined') {
+            publicAPI.setDate();
+        }        
+    }
+    
+    // Function returns date object regardless if argument passed is epoch ms or Date obj
+    function getDateObj(date) {
+        if (date instanceof Date) {
+            return date;
+        } else {
+            return new Date(date);
+        }        
+    }
+
+    // Function returns epoch milliseconds regardless if argument passed is epoch ms or Date obj
+    function getDateMilli(date) {
+        if (date instanceof Date) {
+            return date.getTime();
+        } else {
+            return date;
+        }        
+    }
+    
+    var publicAPI = {
+        
+        // Store date passed in as Date object or epoch milliseconds (default to now)
+        setDate: function setDate(date) {
+            
+            // If no argument passed set date to now 
+            if (typeof date === 'undefined') {
+                date = Date.now();
+                
+            //  If date is NOT a Date obj OR a number (ms), set date to now
+            } else if (!(date instanceof Date) && (typeof date !== 'number')) {
+                date = Date.now();
+            }
+            
+            theDate = date;
+        },
+        
+        // Return date as Date object if argument is true, otherwise return milliseconds
+        getDate: function getDate(isDateObj) {
+            verifyDate();            
+
+            if (isDateObj) {
+                return getDateObj(theDate);
+            } else {
+                return getDateMilli(theDate);
+            }            
+        },
+        
+        // Return day of week ('Monday', etc.)
+        getDayName: function getDayName() {
+            var dayValue;
+            var dayLabel;
+
+            verifyDate();
+            
+            // Use getDay method on Date obj (getDateObj fcn converts from ms to obj if needed)
+            dayValue = getDateObj(theDate).getDay();
+            
+            // Assign weekday label based on weekday value
+            switch (dayValue) {
```

Comment:
Hi James, 

Thanks so much for the detailed comments - I really appreciate the input. I think your hint was referring to using arrays for the weekday & month labels, and I'm not sure what "externalize configuration data" means. Can you please let me know if either of these are on the right track? Thank you! (I see that .toLocaleDateString is another option that doesn't yet work in all browsers.) 

1) Just use var weekdayLabels = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday']; inside my getDayName function and return weekdayLabels[dayValue]; . (Same for getMonthName). 

2) OR, declare an object var outside of my IIFE (instead of inside??) that looks something like this (and then reference it inside my IIFE with dateLabels.getWeekdayLabel(dayValue) ): 



var dateLabels = { 
getWeekdayLabel: function getWeekdayLabel(weekday) { 
var weekdayLabels = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 
'Friday', 'Saturday']; 

return weekdayLabels[weekday]; 
}, 

getMonthLabel: function getMonthLabel(month) { 
var monthLabels = ['January', 'February', 'March', 'April', 'May', 'June', 
'July', 'August', 'September', 'October', 'November', 'December']; 

return monthLabels[month]; 
} 
}; 




#2 above worked for me, but I wanted to check to see if that was what you had in mind. Partly, I'm not sure what it means to externalize data (put it in a separate .js file, or put it in the same .js file outside of the IIFE?). And would we externalize something like this just because it's generic data that might be referenced by a variety of other functions, or is it more efficient to define the dateLabels object outside of the IIFE? 




Thank you! 

Diane 
----- Original Message -----

From: "jamdon" <notifications@github.com> 
To: "JSCRIPT300-Spring2015/Module_2" <Module_2@noreply.github.com> 
Cc: "dianezev" <dianezev@comcast.net> 
Sent: Saturday, April 11, 2015 6:18:07 PM 
Subject: Re: [Module_2] Module 2 HW Diane Zevenbergen (#2) 



In app/js/enhancedDate.js : 
> +                return getDateMilli(theDate);
> +            }            
> +        },
> +        
> +        // Return day of week ('Monday', etc.)
> +        getDayName: function getDayName() {
> +            var dayValue;
> +            var dayLabel;
> +
> +            verifyDate();
> +            
> +            // Use getDay method on Date obj (getDateObj fcn converts from ms to obj if needed)
> +            dayValue = getDateObj(theDate).getDay();
> +            
> +            // Assign weekday label based on weekday value
> +            switch (dayValue) { 


Functions works as written, so my only feedback is just an implementation optimization. Notice how the getDay and getMonth return a number starting with 0? Think of how you can represent the day names and month names as values starting with 0 (hint: there's a particular data structure this maps very well to.) This would allow you to externalize configuration data like this as part of the private data of your module and the methods would have access to that data when returning the specific day or month string respectively. 

— 
Reply to this email directly or view it on GitHub .